src/lib/field.ts

```diff
@@ -0,0 +1,794 @@
+};
+
+type ConstantField = Field & { value: ConstantFieldVar };
+
+/**
+ * An element of a finite field.
+ */
+class Field {
+  value: FieldVar;
+
+  /**
+   * The field order as a `bigint`.
+   */
+  static ORDER = Fp.modulus;
+
+  /**
+   * Coerces anything field-like to a {@link Field}.
+   */
+  constructor(x: bigint | number | string | Field | FieldVar) {
+    if (Field.#isField(x)) {
+      this.value = x.value;
+      return;
+    }
+    // fieldVar
+    if (Array.isArray(x)) {
+      this.value = x;
+      return;
+    }
+    // TODO this should handle common values efficiently by reading from a lookup table
+    this.value = FieldVar.constant(Fp(x));
+  }
+
+  // helpers
+  static #isField(x: bigint | number | string | Field | FieldVar): x is Field {
+    return x instanceof Field || (x as any) instanceof SnarkyFieldConstructor;
+  }
+  static #toConst(x: bigint | number | string | ConstantField): FieldConst {
+    if (Field.#isField(x)) return x.value[1];
+    return FieldConst.fromBigint(Fp(x));
+  }
+  static #toVar(x: bigint | number | string | Field): FieldVar {
+    if (Field.#isField(x)) return x.value;
+    return FieldVar.constant(Fp(x));
+  }
+  static from(x: bigint | number | string | Field): Field {
+    if (Field.#isField(x)) return x;
+    return new Field(x);
+  }
+
+  /**
+   * Checks whether this is a hard-coded constant in the Circuit.
+   */
+  isConstant(): this is { value: ConstantFieldVar } {
+    return this.value[0] === FieldType.Constant;
+  }
+
+  /**
+   * Returns a constant.
+   */
+  toConstant(): ConstantField {
+    if (this.isConstant()) return this;
+    // TODO: fix OCaml error message, `Can't evaluate prover code outside an as_prover block`
+    let value = Snarky.field.readVar(this.value);
+    return new Field(FieldVar.constant(value)) as ConstantField;
+  }
+
+  /**
+   * Serialize this instance of a {@link Field} to a bigint.
+   * This operation does NOT affect the circuit and can't be used to prove anything about the bigint representation of the Field.
+   */
+  toBigInt() {
+    let x = this.toConstant();
+    return FieldConst.toBigint(x.value[1]);
+  }
+
+  /**
+   * Serialize the {@link Field} to a string, e.g. for printing.
+   * This operation does NOT affect the circuit and can't be used to prove anything about the string representation of the Field.
+   */
+  toString() {
+    return this.toBigInt().toString();
+  }
+
+  /**
+   * Assert that this {@link Field} equals another Field-like value.
+   * Throws an error if the assertion fails.
+   *
+   * ```ts
+   * Field(1).assertEquals(1);
+   * ```
+   */
+  assertEquals(y: Field | bigint | number | string, message?: string) {
+    try {
+      if (this.isConstant() && isConstant(y)) {
+        if (this.toBigInt() !== toFp(y)) {
+          throw Error(`Field.assertEquals(): ${this} != ${y}`);
+        }
+        return;
+      }
+      Snarky.field.assertEqual(this.value, Field.#toVar(y));
```

Comment:
am I correct in my assumption that `Snarky` only deals with variables (in-circuit variables?) and everything that is constant related has to be dealt with in JS now?

Reply:
snarky has methods that deal with both, but we don't use them because handling all constant arithmetic on the JS side is an important milestone of the whole refactor. it's the main step towards `snarkyjs/light`